typescript/token/src/adapters/EvmTokenAdapter.ts

```diff
@@ -0,0 +1,213 @@
+  }
+}
+
+// Interacts with Hyp Synthetic token contracts (aka 'HypTokens')
+export class EvmHypSyntheticAdapter<T extends HypERC20 = HypERC20>
+  extends EvmTokenAdapter<T>
+  implements IHypTokenAdapter
+{
+  constructor(
+    public readonly chainName: ChainName,
+    public readonly multiProvider: MultiProtocolProvider,
+    public readonly addresses: { token: Address },
+    public readonly contractFactory: any = HypERC20__factory,
+  ) {
+    super(chainName, multiProvider, addresses, contractFactory);
+  }
+
+  getDomains(): Promise<Domain[]> {
+    return this.contract.domains();
+  }
+
+  async getRouterAddress(domain: Domain): Promise<Buffer> {
+    const routerAddressesAsBytes32 = await this.contract.routers(domain);
+    // Evm addresses will be padded with 12 bytes
+    if (routerAddressesAsBytes32.startsWith('0x000000000000000000000000')) {
+      return Buffer.from(
+        strip0x(bytes32ToAddress(routerAddressesAsBytes32)),
+        'hex',
+      );
+      // Otherwise leave the address unchanged
+    } else {
+      return Buffer.from(strip0x(routerAddressesAsBytes32), 'hex');
+    }
+  }
+
+  async getAllRouters(): Promise<Array<{ domain: Domain; address: Buffer }>> {
+    const domains = await this.getDomains();
+    const routers: Buffer[] = await Promise.all(
+      domains.map((d) => this.getRouterAddress(d)),
+    );
+    return domains.map((d, i) => ({ domain: d, address: routers[i] }));
+  }
+
+  async quoteGasPayment(destination: Domain): Promise<string> {
+    const gasPayment = await this.contract.quoteGasPayment(destination);
+    return gasPayment.toString();
+  }
+
+  populateTransferRemoteTx({
+    weiAmountOrId,
+    destination,
+    recipient,
+    txValue,
+  }: TransferRemoteParams): Promise<PopulatedTransaction> {
+    const recipBytes32 = addressToBytes32(addressToByteHexString(recipient));
+    return this.contract.populateTransaction.transferRemote(
+      destination,
+      recipBytes32,
+      weiAmountOrId,
+      {
+        value: txValue,
+      },
+    );
+  }
+}
+
+// Interacts with HypCollateral and HypNative contracts
+export class EvmHypCollateralAdapter
+  extends EvmHypSyntheticAdapter
+  implements IHypTokenAdapter
+{
+  constructor(
+    public readonly chainName: ChainName,
+    public readonly multiProvider: MultiProtocolProvider,
+    public readonly addresses: { token: Address },
+    public readonly contractFactory: any = HypERC20Collateral__factory,
+  ) {
+    super(chainName, multiProvider, addresses, contractFactory);
+  }
+
+  override getMetadata(): Promise<MinimalTokenMetadata> {
+    // TODO pass through metadata from wrapped token or chainMetadata config
+    throw new Error(
+      'Metadata not available for HypCollateral/HypNative contract.',
+    );
+  }
+
+  override populateApproveTx(
+    _params: TransferParams,
+  ): Promise<PopulatedTransaction> {
+    throw new Error(
+      'Approve not applicable to HypCollateral/HypNative contract.',
```

Comment:
maybe
`collateral.approve(router, amount)`?

Reply:
Sorry I don't follow, can you clarify?